src/mqt/predictor/reward.py

```diff
@@ -69,3 +70,90 @@ def calc_qubit_index(qargs: list[Qubit], qregs: list[QuantumRegister], index: in
             return qubit_index
     error_msg = f"Global qubit index for local qubit {index} index not found."
     raise ValueError(error_msg)
+
+
+def expected_success_probability(qc: QuantumCircuit, device: Device, precision: int = 10) -> float:
+    """Calculates the expected success probability of a given quantum circuit on a given device.
+
+    Args:
+        qc (QuantumCircuit): The quantum circuit to be compiled.
+        device(mqt.bench.Device): The device to be used for compilation.
+        precision (int, optional): The precision of the returned value. Defaults to 10.
+
+    Returns:
+        float: The expected success probability of the given quantum circuit on the given device.
+    """
+    res = 1.0
+
+    # collect gate durations for all qubits
+    op_times = []
+    for op in device.basis_gates:
+        if op == "barrier" or op == "id":
+            continue
+        if op in ["rxx", "rzz", "cx", "cz", "cp", "ecr", "xx_plus_yy"]:
+            for q0, q1 in device.coupling_map:  # multi-qubit gates
+                try:
+                    d = device.get_two_qubit_gate_duration(op, q0, q1)
+                    op_times.append((op, [q0, q1], d, "s"))
+                except ValueError:  # noqa:PERF203
+                    pass  # gate duration not available for this qubit pair
+        else:  # single-qubit gates
+            for q in range(device.num_qubits):
+                try:
+                    if op == "measure":
+                        d = device.get_readout_duration(q)
+                    else:
+                        d = device.get_single_qubit_gate_duration(op, q)
+                    op_times.append((op, [q], d, "s"))
+                except ValueError:  # noqa:PERF203
+                    pass  # gate duration not available for this qubit
+
+    # associate gate times for each qubit through asap scheduling
+    transpiled = transpile(qc, scheduling_method="asap", basis_gates=device.basis_gates, instruction_durations=op_times)
+
+    qubit_durations: dict[int, list[float]] = {qubit_index: [] for qubit_index in range(device.num_qubits)}
+
+    for instruction, qargs, _cargs in transpiled.data:
+        gate_type = instruction.name
+        if gate_type == "barrier":
+            continue
+
+        assert len(qargs) in [1, 2]
+        first_qubit_idx = calc_qubit_index(qargs, qc.qregs, 0)
+        if device.name == "rigetti_aspen_m2":
```

Comment:
Is it possible to avoid special treatment of certain devices and ideally include it in the respective device/provider in MQT Bench or, if that does not work, do it in the `calc_qubit_index` function?